packages/cli-vector/src/generalization/generalization.ts

```diff
@@ -32,87 +30,118 @@ export async function generalize(input: URL, output: URL, options: VectorCreatio
+      outputCount++;
     }
-    writeStream.close();
-    return output;
   }
-  return undefined;
+
+  await new Promise((resolve) => {
+    writeStream.close(resolve);
+  });
+
+  // special handling for the gazatteer (place_labels) dataset
+  // issues: bypasses the 'simplify' and 'removeAttributes' operations of the 'tag' function
+  // if (options.layer.id === '51154') {
+  //   logger.info({}, 'Special handling for the gazatteer (place_labels) dataset');
+  //   features = Array.from(PlaceLabelsFeatures.values());
+  //   outputCount = features.length;
+  // }
+
+  const metrics: Metrics = {
+    input: inputCount,
+    output: outputCount,
+  };
+
+  logger.info({ inputCount, outputCount }, 'Generalize:End');
+  return metrics;
 }
 
 /**
  * Tag feature for layer
  */
-function tag(options: VectorCreationOptions, line: string, simplify?: Simplify): string | undefined {
-  const feature = JSON.parse(line) as VectorGeoFeature;
-
-  feature['tippecanoe'] = {
-    layer: options.name,
-    minzoom: options.layer.style.minZoom,
-    maxzoom: options.layer.style.maxZoom,
-  };
+function tag(
+  options: VectorCreationOptions,
+  line: string,
+  simplify: Simplify | null,
+  logger: LogType,
+): VectorGeoFeature | null {
+  logger.debug({}, 'Tag:Start');
+  const feature = VectorGeoFeatureSchema.parse({
+    ...JSON.parse(line),
+    tippecanoe: {
+      layer: options.name,
+      minzoom: options.layer.style.minZoom,
+      maxzoom: options.layer.style.maxZoom,
+    },
+  });
 
+  // copy the stac json's tags to the feature (i.e. 'kind')
   Object.entries(options.layer.tags).forEach(([key, value]) => (feature.properties[key] = value));
 
-  // TODO: Add support of special tagging logics.
+  // adjust the feature's metadata and properties
+  const modifiedFeature = modifyFeature(feature, options, logger);
+  if (modifiedFeature == null) {
+    logger.debug({}, 'Tag:End');
+    return null;
+  }
 
   // Simplify geometry
   if (simplify != null) {
     // Update the simplified feature zoom level
-    feature['tippecanoe'] = {
+    modifiedFeature['tippecanoe'] = {
       layer: options.name,
       minzoom: simplify.style.minZoom,
       maxzoom: simplify.style.maxZoom,
     };
     if (simplify.tolerance != null) {
-      const geom = feature.geometry;
+      const geom = modifiedFeature.geometry;
       const type = geom.type;
       const coordinates = simplifyFeature(type, geom, simplify.tolerance);
-      if (coordinates == null) return undefined;
-      feature.geometry = coordinates;
+      if (coordinates == null) {
+        logger.debug({}, 'Tag:End');
+        return null;
+      }
+      modifiedFeature.geometry = coordinates;
     }
   }
 
   // Remove unused properties
-  removeAttributes(feature, options);
+  // REVIEW: this function just removes the special tags. something isn't right here
+  const cleanedFeature = removeAttributes(modifiedFeature, options);
 
-  return JSON.stringify(feature);
+  logger.debug({}, 'Tag:End');
```

Comment:
Same dumping logs too

Reply:
I believe we want the logs like `ndjson start/end`, `generaize start/end` and `mbtiles start/end` for each layer.